presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -299,89 +296,46 @@ public DictionaryBlock compact()
             return this;
         }
 
-        List<Block> compactBlocks = compactBlocks(singletonList(this));
-        return (DictionaryBlock) compactBlocks.get(0);
-    }
-
-    public static List<Block> compactBlocks(List<Block> blocks)
-    {
-        verifyEligibleToCompact(blocks);
-
-        DictionaryBlock dictionaryBlock = (DictionaryBlock) blocks.get(0);
-        Block dictionary = dictionaryBlock.getDictionary();
-        Slice ids = dictionaryBlock.getIds();
-
-        int positionCount = dictionaryBlock.getPositionCount();
+        // determine which dictionary entries are referenced and build a reindex for them
         int dictionarySize = dictionary.getPositionCount();
-
         List<Integer> dictionaryPositionsToCopy = new ArrayList<>(dictionarySize);
         int[] remapIndex = new int[dictionarySize];
         Arrays.fill(remapIndex, -1);
-        int newIndex = 0;
 
-        boolean[] copied = new boolean[dictionarySize];
-        for (int i = 0; i < positionCount; i++) {
-            int position = getIndex(ids, i);
-            if (!copied[position]) {
-                dictionaryPositionsToCopy.add(position);
-                remapIndex[position] = newIndex;
+        boolean[] isReferenced = new boolean[dictionarySize];
```

Comment:
You can replace `isReferenced ` with a comparison for equality with -1 in `remapIndex`